### PR TITLE
[FEATURE] Add thread-local logging to fix data races (issue #8596)

### PR DESCRIPTION
**Pull request by timosachsenberg**
This adds thread-local LogStream instances to eliminate data races in the logging system. The race conditions occur because sync() modifies shared state (static buffer, incomplete_line_, log_cache_) without protection.

Key changes:
- Add ThreadLogContext class with thread-local storage for log streams
- Add OPENMS_LOG_*_TL macros for direct thread-local access
- Add OPENMS_THREADLOCAL_LOGGING CMake option (enabled by default)
- When enabled, OPENMS_LOG_* macros use thread-local streams
- Update LogStream_test to work with both modes
- Add comprehensive ThreadLogContext_test

Each thread now gets its own LogStream instances, eliminating all races:
- static char buf[] - each thread has own buffer
- incomplete_line_ - each thread has own string
- log_cache_/log_time_cache_ - each thread has own caches

Users can disable with: cmake -DOPENMS_THREADLOCAL_LOGGING=OFF ..

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

## Summary by CodeRabbit

* **New Features**
  * Added thread-local logging support with dedicated per-thread log stream instances.
  * Introduced new flushIncomplete() methods to properly handle buffered incomplete lines.
  * Enhanced logging configuration via new public stream accessors and exports.

* **Tests**
  * Expanded test coverage for thread-local and global stream logging behavior.

* **Chores**
  * Updated logging macros to support both thread-local and legacy logging paths.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by timosachsenberg** [comment]
once confirmed correct will remove flag

File: src/openms/include/OpenMS/CONCEPT/LogStream.h [hunk 1]
```diff
@@ -433,36 +434,109 @@ namespace OpenMS
...
+
+  /// Legacy macro for fatal errors - has data race bugs, use OPENMS_LOG_FATAL_ERROR_TL instead
+#define OPENMS_LOG_FATAL_ERROR_LEGACY \
   OPENMS_THREAD_CRITICAL(LOGSTREAM) \
   OpenMS_Log_fatal << __FILE__ << "(" << __LINE__ << "): "
 
-  /// Macro to be used if non-fatal error are reported (processing continues)
-#define OPENMS_LOG_ERROR \
+  /// Legacy macro for non-fatal errors - has data race bugs
+#define OPENMS_LOG_ERROR_LEGACY \
```
Comment (cbielow) [inline src/openms/include/OpenMS/CONCEPT/LogStream.h]: mhh... a simply function which avoids exposing a thread_local variable to the ABI would have done the trick, no?!
```
Logger::LogStream& getOpenMSLogFatal() { thread_local Logger::LogStream tls_log; return tls_log; }
```
??
The solution from Claude is more flexible, but also horribly overengineered?!

Comment (timosachsenberg) [reply src/openms/include/OpenMS/CONCEPT/LogStream.h]: I think this is currently because it wanted to pertain API / backwards compatibility with some other classees. If this is going in the right direction I will completely switch to this approach and simplify.

Comment (timosachsenberg) [reply src/openms/include/OpenMS/CONCEPT/LogStream.h]: Addressing cbielow's Concern

  The reviewer suggested:
  Logger::LogStream& getOpenMSLogFatal() {
      thread_local Logger::LogStream tls_log;
      return tls_log;
  }

  Why the current design is more complex:
  - Thread-local streams need to inherit configuration from somewhere (the globals)
  - flushAll() is needed before LogConfigHandler reconfigures log levels
  - reset() is needed for test cleanup

  Simplest viable approach (if we keep configuration inheritance):
  // Could replace ThreadLogContext class with 5 simple functions:
  Logger::LogStream& getThreadLocalFatal() {
      thread_local Logger::LogStream tls(OpenMS_Log_fatal);  // Copy from global
      return tls;
  }
  // ... repeat for error, warn, info, debug

  But this loses:
  - Coordinated flushAll() before reconfiguration
  - reset() for test cleanup
  - isInitialized() checks

  Recommendation

  Two options to present to cbielow:

  1. Keep ThreadLogContext but remove all legacy code paths (~135 lines deleted, class stays)
  2. Replace with simple functions but then need to:
    - Remove flushAll() call from LogConfigHandler (accept potential partial-line issues)
    - Handle test cleanup differently (or accept state leakage between tests)

  Want me to implement option 1 (remove legacy code, keep ThreadLogContext), or prepare a prototype of option 2 (simple functions)?

**Comment by timosachsenberg** [comment]
## Response to Over-Engineering Concern

@cbielow You raised a good point about whether a simpler approach would suffice:

```cpp
Logger::LogStream& getOpenMSLogFatal() { 
    thread_local Logger::LogStream tls_log; 
    return tls_log; 
}
```

The `ThreadLogContext` class exists primarily for three functions that simple standalone functions can't provide:

### 1. `flushAll()` - Called Before Log Reconfiguration

Used in `LogConfigHandler.cpp` before reconfiguring log streams. Flushes incomplete lines (text without `\n`) from ALL thread-local buffers.

**Without it:** If a thread is mid-line when `LogConfigHandler` reconfigures streams, partial output could be lost or garbled. This is a rare edge case though.

### 2. `reset()` - Test Isolation (Most Important)

Used in `LogStream_test.cpp`, `LogConfigHandler_test.cpp`, and `Param_test.cpp` to destroy thread-local instances between tests.

**Without it:** Test pollution - one test's log configuration leaks into subsequent tests, causing flaky/non-deterministic test results. Tests redirect log streams to capture output, and without reset, those redirects persist.

### 3. `isInitialized()` / `ensureInitialized()` - Lifecycle Management

Ensures thread-local streams copy the *current* global configuration when created.

**Without it:** Edge cases where configuration timing matters might behave unexpectedly.

---

### Summary

| Function | If Lost | Severity |
|----------|---------|----------|
| `flushAll()` | Rare garbled output during reconfiguration | Low |
| `reset()` | Test pollution, flaky tests | **High** |
| `isInitialized()`/`ensureInitialized()` | Config timing edge cases | Low |

The class could be replaced with 5 simple functions **if** we accept losing `reset()` (and accept test isolation issues) or implement an alternative reset mechanism.

**Options:**
1. Keep `ThreadLogContext` as-is (current approach)
2. Simplify to functions + add a `thread_local bool` reset flag for tests
3. Simplify to functions and accept test isolation challenges

Let me know your preference and I can refactor accordingly.

**Comment by cbielow** [comment]
these are good points, let me address them:

```
1. flushAll() - Called Before Log Reconfiguration
Used in LogConfigHandler.cpp before reconfiguring log streams. Flushes incomplete lines (text without \n) from ALL thread-local buffers.
Without it: If a thread is mid-line when LogConfigHandler reconfigures streams, partial output could be lost or garbled. This is a rare edge case though.
```
This should not be a supported use-case. Log configuration should be done once in the beginning globally, and then copied by all threads.

```
2. reset() - Test Isolation (Most Important)

Used in LogStream_test.cpp, LogConfigHandler_test.cpp, and Param_test.cpp to destroy thread-local instances between tests.

Without it: Test pollution - one test's log configuration leaks into subsequent tests, causing flaky/non-deterministic test results. Tests redirect log streams to capture output, and without reset, those redirects persist.
3. isInitialized() / ensureInitialized() - Lifecycle Management

Ensures thread-local streams copy the current global configuration when created.
```
What prevents you from resetting the thread_local streams in the tests?
Also, consider that a thread-pool, used by OpenMP and others, will re-use the same thread_local variable across different parallel regions (just tested this manually), i.e. one cannot expect a new vanilla thread in a parallel region!

**Comment by cbielow** [comment]
the loghandler is complicated enough as it is. The new structure adds too much complexity and increases maintenance burden (since wrapping your head around it and reasoning about it is very hard). So keep it as simple as possible and document restrictions and do's and don'ts.

**Comment by cbielow** [comment]
I am almost inclined to go back to a single shared global state with mutexes, although this means a potentially considerable performance hit (we should actually measure this for tools that are heavy in logging).

**Comment by timosachsenberg** [comment]
Thanks for testing OpenMP. starting to understand why some logging libraries took years to stabilize :)

**Comment by jpfeuffer** [comment]
How about using a logging library then? Do we really need this weird log line deduplication functionality that makes multithreading even more complex? We could have Boost.log "for free".
But yeah I'd rather fix it here if not too complex.

**Comment by timosachsenberg** [comment]
## Simplification Complete

Based on @cbielow's feedback, I've simplified the implementation from `ThreadLogContext` class to 5 simple functions:

```cpp
Logger::LogStream& getThreadLocalLogFatal()
{
    thread_local Logger::LogStream tls(new Logger::LogStreamBuf(OpenMS_Log_fatal.rdbuf(), &red), true);
    return tls;
}
// ... same pattern for error, warn, info, debug
```

### What changed:
- **Removed** `ThreadLogContext` class entirely (~250 lines)
- **Removed** `OPENMS_THREADLOCAL_LOGGING` CMake feature flag
- **Removed** legacy macros - thread-local is now the only implementation
- **Added** 5 simple accessor functions (~35 lines)
- **Net change:** -660 lines

### Documented restrictions (in LogStream.h):
- Log configuration should be done ONCE at program start, BEFORE spawning threads
- Thread-local streams copy config from globals on first access
- OpenMP thread pools reuse threads, so thread_local state persists across parallel regions

### Test results:
- ✅ LogStream_test: Passed
- ✅ LogConfigHandler_test: Passed (updated to test global streams directly)
- ✅ 99% of all tests pass (2428/2437) - failures unrelated to logging

### Note on performance:
The main benefit is **correctness** (fixing data races from issue #8596), not speed. With shared stdout, writes still serialize at the OS level. Each thread can now buffer in parallel, but the real win is eliminating undefined behavior from the original race conditions on `static char buf[]`, `incomplete_line_`, and cache maps.

**Comment by jpfeuffer** [comment]
Would be interesting to see what happens if you merge this in the PR with the segfaults.
Or would you want to merge it anyway?
Have you compared the testing times for example?

Test looks similar (slightly faster if not noise).

**Comment by timosachsenberg** [comment]
I can try that. No strong opinion on merging or not. Interestingly I had a bug in this PR that als triggered a segfault in Param_test. So there is a Chance …

**Comment by timosachsenberg** [comment]
Superseded
